### ISSUE TITLE
file provisioner is unable to download directory

### ISSUE THREAD

**Issue by ltrifonov** [mitchellh/packer#3251]
Tried to download a folder form the remote instance locally.
Tested all possible combinations `/` on  source/destinations with no luck.
Quick look at the code showed that the code checks for directory on uploads only.

The documentation doesn't mention that folder download is not possible?
 
I've setup my file provisioner like:
```javascript
{
    "type": "file", 
    "source": "/tmp/htmlreports/", 
    "direction": "download",
    "destination": "htmlreports/" 
},
```
The provisioner always fails with message
```
==> amazon-ebs: Downloading /tmp/htmlreports/ => htmlreports/
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: No AMIs to cleanup
==> amazon-ebs: Deleting temporary security group...
Build 'amazon-ebs' errored: open htmlreports/: is a directory

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: open htmlreports/: is a directory

==> Builds finished but no artifacts were created.
```

**Comment by vtolstov** [comment]
Try source without / at the end
23 февр. 2016 г. 10:02 пользователь "Lyubomir Trifonov" <
notifications@github.com> написал:

**Comment by ltrifonov** [comment]
Thanks for the suggestion, already tried. 
To be more specific on what I've tried:

destination without slash:
```
==> amazon-ebs: Downloading /tmp/htmlreports/ => htmlreports
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: No AMIs to cleanup
==> amazon-ebs: Deleting temporary security group...
Build 'amazon-ebs' errored: open htmlreports: is a directory

==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: open htmlreports: is a directory
```

As suggested: no slash at the source end:
```
==> amazon-ebs: Downloading /tmp/htmlreports => htmlreports
==> amazon-ebs: Download failed: scp: /tmp/htmlreports: not a regular file
==> amazon-ebs:
==> amazon-ebs: Terminating the source AWS instance...
==> amazon-ebs: No AMIs to cleanup
==> amazon-ebs: Deleting temporary security group...
Build 'amazon-ebs' errored: scp: /tmp/htmlreports: not a regular file


==> Some builds didn't complete successfully and had errors:
--> amazon-ebs: scp: /tmp/htmlreports: not a regular file

```

**Comment by rickard-von-essen** [comment]
@ltrifonov Could you rerun with packer log ```PACKER_LOG=1 packer build ...``` and gist the log.

**Comment by cbednarski** [comment]
@ltrifonov Downloading directories is only supported in when using SSH. If you're using Docker or Windows or something else this will not work (it's not currently implemented).

It's possible that the error message is being swallowed. Can you indicate which builder / communicator you're using? The file provisioner doesn't actually know how to talk to the machine so it relies on the communicator to do this.

**Comment by rickard-von-essen** [comment]
```
Build 'amazon-ebs' errored: scp: /tmp/htmlreports: not a regular file
```
Suggests SSH.

**Comment by vtolstov** [comment]
Im try to recheck this with qemu.
23 февр. 2016 г. 12:40 пользователь "Rickard von Essen" <
notifications@github.com> написал:

**Comment by iroller** [comment]
Having the same issue here.

**Comment by mohae** [comment]
@cbednarski this issue touches on the topic I was hinting at when I said I'd probably open a separate issue about the docs  in my pull request that addressed #3474.  Instead, I think this is an appropriate place to bring it up.

The file provisioner docs only talk about file uploads and directory handling.  I was thinking that the download direction should be handled in the same manner with the docs being updated so that the "Directory Uploads" section.

As currently implemented the file provisioner's download only supports a file.  The short-circuit code that is in `ProvisionUpload` for handling directories is not part of the `ProvisionDownload`.  There is a `packer.Communicator.DownloadDir` method.

The resolution would be to add the `DownloadDir` short circuit to `ProvisionDownload` and update the documentation with what I suggested above along with updating it to note that the File provisioner will create all the directories in the specified destination path (which would make it consistent with the changes in https://github.com/mitchellh/packer/pull/3692.

I'd be happy to do the described changes.

**Comment by cbednarski** [comment]
Yeah I think this needs to be a docs fix for now. The problem with downloading directories is we need a way to enumerate (`tree`) the remote filesystem across different communicators and operating systems, and we don't currently have this. It's a non-trivial addition. `rsync` accomplishes this by running a pid on the remote side.

**Comment by vtolstov** [comment]
2016-07-08 22:31 GMT+03:00 Chris Bednarski <notifications@github.com>:

> Yeah I think this needs to be a docs fix for now. The problem with
> downloading directories is we need a way to enumerate (tree) the remote
> filesystem across different communicators and operating systems, and we
> don't currently have this. It's a non-trivial addition. rsync
> accomplishes this by running a pid on the remote side.
>

I think that download dir is possible, but need some work. If i find docs i
can take this.


-- 
Vasiliy Tolstov,
e-mail: v.tolstov@yoctocloud.net

**Comment by Ablu** [comment]
> I think that download dir is possible, but need some work. If i find docs i can take this.

Hm... So does it work somehow? If so how? I seem to hit the same bug (with OpenStack + SSH).

**Comment by cbednarski** [comment]
Hey folks, this does not currently work unless you use an out of band process like `scp` to copy files. The reason is we don't have a cross-platform, cross-communicator way to do a remote, recursive `ls`. When we do directory uploads we can use local filesystem enumeration to figure out what to copy but typically `ssh` or `rsync` does this by running pids on both ends of the copy so we can't use the same trick in Go.

I'm changing this to an enhancement / docs since this is not something we have implemented yet. I understand for many of you this seems like a bug, but it's currently missing functionality. I don't want to the tags to imply that this might work somehow and you are doing it wrong.

**Comment by vtolstov** [comment]
scp allows to download dir, i found somem spec: https://blogs.oracle.com/janp/entry/how_the_scp_protocol_works

**Comment by mwhooker** [comment]
If you're using scp to download the directory, this will be fixed via #3899

This is in master now, but should make it to 0.11

please let me know if you continue to see issues
